lib/grp.gi

```diff
@@ -4291,86 +4294,121 @@ end );
 InstallMethod( GroupWithGenerators,
     "generic method for collection",
     [ IsCollection ],
-    function( gens )
-    local G,fam,typ;
-
-    fam:=FamilyObj(gens);
-    if IsFinite(gens) then
-      if not IsBound(fam!.defaultFinitelyGeneratedGroupType) then
-        fam!.defaultFinitelyGeneratedGroupType:=
-          NewType(fam,IsGroup and IsAttributeStoringRep
-                      and HasGeneratorsOfMagmaWithInverses
-                      and IsFinitelyGeneratedGroup);
-      fi;
-      typ:=fam!.defaultFinitelyGeneratedGroupType;
-    else
-      if not IsBound(fam!.defaultGroupType) then
-        fam!.defaultGroupType:=
-          NewType(fam,IsGroup and IsAttributeStoringRep
-                      and HasGeneratorsOfMagmaWithInverses);
+function( gens )
+local G,fam,typ,id,triv;
+
+  fam:=FamilyObj(gens);
+
+  triv:=fail; # might not find out
+
+  typ:=IsGroup and IsAttributeStoringRep 
+    and HasIsEmpty and HasGeneratorsOfMagmaWithInverses;
+  if IsFinite(gens) then
+    typ:=typ and IsFinitelyGeneratedGroup;
+    if Length(gens)<=1 then
+      typ:=typ and HasIsCyclic and IsCyclic 
+          and HasIsCommutative and IsCommutative;
+    fi;
+
+    if Length(gens)>0 and CanEasilyCompareElements(gens) then
+      id:=One(gens[1]);
+      if ForAny(gens,x->x<>id) then
+        typ:=typ and HasIsTrivial and HasIsNonTrivial;
+        triv:=false;
+      else
+        typ:=typ and HasIsTrivial and HasIsNonTrivial;
+        triv:=true;
+        # basic consequence
+        typ:=typ and HasIsPerfectGroup and IsPerfectGroup;
       fi;
-      typ:=fam!.defaultGroupType;
     fi;
 
-    G:=rec();
-    ObjectifyWithAttributes(G,typ,GeneratorsOfMagmaWithInverses,AsList(gens));
+  fi;
+  typ:=NewType(fam,typ);
```

Comment:
Just to make it explicit: the changes in this PR get rid of the cached values `fam!.defaultFinitelyGeneratedGroupType` and `fam!.defaultGroupType`. I assume that you verified that there is a net gain (probably because this initial type (almost?) always ended up being replaced by immediate methods anyway.

Reply:
When I left the caching in I ran into problems -- a non-nilpotent group claimed to be nilpotent. The best I could come up (#2394 ) is that the cached type in fact gets changed later and thus cannot be cached. If I'm doing something wrong here I'd love to know (and then would add type caching again).

Reply:
I am not sure how useful those caches would be, unless done very carefully: in the old code, there were already two caches, one for finitely generated groups, and one for those where we don't know whether they are f.g. or not. In order to preset `IsTrivial` and other flags, you'd have to either create more caches (i.e. have one cached type for trivial groups; one for non-trivial cyclic; one for non-cyclic f.g.; one for the generic case). Of course that number grows as you add special cases. Luckily, in this case at least the number of types does not seem to grow exponentially, as most of the checks differentiate based on the generating set along.

Reply:
At the start of writing the previous comment, I wasn't sure, because I thought you'd need to add 2^n caches, for n 3 or 4, but now it seems only four may be needed, which is at least not totally crazy.

Reply:
Another cause for the problems you have seen may be that you add e.g. `HasIsTrivial` to the flag lists of the type. I think that means that any object created with that type is *non-trivial*, no matter what; once `HasIsTrivial` is set, one cannot change the value of `IsTrivial` anymore (at least not without low-level bit fiddling); this is a feature, not a bug.